include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/ExtendedTaintAnalysis/AbstractMemoryLocation.h

```diff
@@ -0,0 +1,231 @@
+/******************************************************************************
+ * Copyright (c) 2020 Fabian Schiebel.
+ * All rights reserved. This program and the accompanying materials are made
+ * available under the terms of LICENSE.txt.
+ *
+ * Contributors:
+ *     Fabian Schiebel and others
+ *****************************************************************************/
+
+#ifndef PHASAR_PHASARLLVM_IFDSIDE_PROBLEMS_EXTENDEDTAINTANALYSIS_ABSTRACTMEMORYLOCATION_H_
+#define PHASAR_PHASARLLVM_IFDSIDE_PROBLEMS_EXTENDEDTAINTANALYSIS_ABSTRACTMEMORYLOCATION_H_
+
+#include <cstdint>
+#include <iosfwd>
+#include <memory>
+#include <optional>
+#include <string>
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/DenseMapInfo.h"
+#include "llvm/ADT/FoldingSet.h"
+#include "llvm/ADT/PointerIntPair.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/IR/DataLayout.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "phasar/PhasarLLVM/DataFlowSolver/IfdsIde/LLVMZeroValue.h"
+#include "phasar/PhasarLLVM/Pointer/LLVMPointsToInfo.h"
+#include "phasar/Utils/LLVMShorthands.h"
+
+namespace psr {
+
+namespace detail {
+
+struct AbstractMemoryLoactionStorage : public llvm::FoldingSetNode {
+  const llvm::Value *Baseptr;
+  uint32_t Lifetime;
+  uint32_t NumOffsets;
+  ptrdiff_t Offsets[0]; // NOLINT
```

Comment:
This is a design pattern which allows to inline a dynamic array into a struct. In the [allocator](https://github.com/secure-software-engineering/phasar/blob/f-Taint/lib/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/ExtendedTaintAnalysis/AbstractMemoryLocationFactory.cpp#L61) we are intentionally allocating more bytes than `sizeof(AbstractMemoryLocationImpl)` to fill the `Offsets` array. Due to intentional overindexing, we can use the `Offsets` field for accessing the "extra" bytes.

This pattern has two major advantages:
- It consumes less bytes than `std::vector`
- Due to inlining, we have better cache locality

Originally this was a `std::array`, but since we need to specify the _k_-limit at runtime (e.g. via a CLI parameter), having a compile-time constant _k_ is no longer useful. Moreover, not all `AbstractMemoryLocation`s make use of all _k_ offset-slots, so allocating _k_ slots for each AML is a waste of space.

Reply:
Ah that makes sense.